libyul/AST.h

```diff
@@ -35,56 +35,81 @@ namespace solidity::yul
-struct Block { langutil::SourceLocation location; std::vector<Statement> statements; };
+struct Block { std::shared_ptr<DebugData const> debugData; std::vector<Statement> statements; };
 /// Function definition ("function f(a, b) -> (d, e) { ... }")
-struct FunctionDefinition { langutil::SourceLocation location; YulString name; TypedNameList parameters; TypedNameList returnVariables; Block body; };
+struct FunctionDefinition { std::shared_ptr<DebugData const> debugData; YulString name; TypedNameList parameters; TypedNameList returnVariables; Block body; };
 /// Conditional execution without "else" part.
-struct If { langutil::SourceLocation location; std::unique_ptr<Expression> condition; Block body; };
+struct If { std::shared_ptr<DebugData const> debugData; std::unique_ptr<Expression> condition; Block body; };
 /// Switch case or default case
-struct Case { langutil::SourceLocation location; std::unique_ptr<Literal> value; Block body; };
+struct Case { std::shared_ptr<DebugData const> debugData; std::unique_ptr<Literal> value; Block body; };
 /// Switch statement
-struct Switch { langutil::SourceLocation location; std::unique_ptr<Expression> expression; std::vector<Case> cases; };
-struct ForLoop { langutil::SourceLocation location; Block pre; std::unique_ptr<Expression> condition; Block post; Block body; };
+struct Switch { std::shared_ptr<DebugData const> debugData; std::unique_ptr<Expression> expression; std::vector<Case> cases; };
+struct ForLoop { std::shared_ptr<DebugData const> debugData; Block pre; std::unique_ptr<Expression> condition; Block post; Block body; };
 /// Break statement (valid within for loop)
-struct Break { langutil::SourceLocation location; };
+struct Break { std::shared_ptr<DebugData const> debugData; };
 /// Continue statement (valid within for loop)
-struct Continue { langutil::SourceLocation location; };
+struct Continue { std::shared_ptr<DebugData const> debugData; };
 /// Leave statement (valid within function)
-struct Leave { langutil::SourceLocation location; };
+struct Leave { std::shared_ptr<DebugData const> debugData; };
 
 struct LocationExtractor
 {
 	template <class T> langutil::SourceLocation operator()(T const& _node) const
 	{
-		return _node.location;
+		return _node.debugData ? _node.debugData->location : langutil::SourceLocation{};
 	}
 };
 
-/// Extracts the source location from an inline assembly node.
+/// Extracts the source location from a Yul node.
 template <class T> inline langutil::SourceLocation locationOf(T const& _node)
 {
 	return std::visit(LocationExtractor(), _node);
 }
 
+struct DebugDataExtractor
+{
+	template <class T> std::shared_ptr<DebugData const> const& operator()(T const& _node) const
+	{
+		return _node.debugData;
+	}
+};
+
+/// Extracts the debug data from a Yul node.
+template <class T> inline std::shared_ptr<DebugData const> const& debugDataOf(T const& _node)
+{
+	return std::visit(DebugDataExtractor(), _node);
```

Comment:
Why do we need a visitor here? Wouldn't this work just as well while being much simpler?

```suggestion
	return _node.debugData;
```

Reply:
The think here is that `_node` is `std::variant<...>` and we can't access it directly like that...
Maybe more readable and nice option can be create base structure for all AST structure so they can inherit from that one, and in that case we could create nicer helper here...

Reply:
Right. I missed the fact that `_node` here is `Expression` or `Statement` rather than just one of the structs defined above. Actually, it might be a good idea to add a comment about that because the `variant` definitions are in `ASTForward.h` so this might not be obvious to everyone.

In that case I can see why it was done like this though personally I'd probably just define it as `debugDataOf(std::variant<T> const& _node)` and use `get<T>(_node).debugData` if I was implementing it myself. I'm fine with leaving it as is though.

Reply:
We can change it once this is merged.